Add `prefer-https` rule

himynameisdave opened the pull request:
Adds `prefer-https` rule.

Checks string literals, as well as comments.

Resolves #971

himynameisdave commented:
I set the [`meta.type`](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics) to `'problem'`, not sure if it should be `'suggestion'` instead though.

sindresorhus commented on docs/rules/prefer-https.md:
```diff
@@ -0,0 +1,24 @@
+# Enforce using HTTPS URLs over HTTP
+
+Enforce using HTTPS URLs over HTTP.
```
This should describe in what places this is enforced.

sindresorhus commented on rules/prefer-https.js:
```diff
@@ -0,0 +1,55 @@
+};
+
+//	Helper to determine if a string is an unsafe URL.
+const containsHttp = nodeValue => /http:\/\/(www\.)?[\w#%+.:=@~-]{2,256}\.[a-z]{2,4}\b([\w#%&+./:=?@~-]*)/.test(nodeValue);
```
We don't need to match more after the TLD.

sindresorhus commented on rules/prefer-https.js:
```diff
@@ -0,0 +1,55 @@
+const create = context => {
+	const sourceCode = context.getSourceCode();
+	const reportPreferHttp = node => {
+		const fixed = node.value.replace('http', 'https');
```
What if the comment is `// Foo bar http is good: https://asdasd.com`?

I think just replacing `http` substring is too fragile.

sindresorhus commented on rules/prefer-https.js:
```diff
@@ -0,0 +1,55 @@
+module.exports = {
+	create,
+	meta: {
+		type: 'problem',
```
This is not correct.

sindresorhus commented on test/prefer-https.js:
```diff
@@ -0,0 +1,14 @@
+		'const foo = \'http://sindresorhus.com\'',
+		'new URL(\'http://sindresorhus.com\')'
+	]
+});
```
You need to add a *lot* more tests.

fisker commented:
Is it possible to disable this rule on comments? Just a question, I never disabled problems in comments.

```js
// eslint-disable-next-line
// http://www.example.com
````

himynameisdave commented:
Hi @sindresorhus.

I've addressed your comments but am still struggling with the tests. They are currently failing in CI due to some weird module error: `'import' and 'export' may appear only with 'sourceType: module'` (see [here](https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1084/checks?check_run_id=1943974197)).

In general I found it quite difficult to debug this new rule, specifically debugging the output of the `fix`. Not sure what your prescribed method for debugging this would be, but usually for an eslint rule I would just run `eslint ./some-test-file.js --fix` to understand what's happening, although we don't seem to use eslint in that way for `run-rules-on-codebase`.

Another roadblock that I have run into is that, because of the nature of this rule, I needed to disable the rule on the test (because the test cases were being flagged when running `run-rules-on-codebase`). However when I then went to run `npm test`, it complains that there is no definition for the rule!?

All this to say is that I am happy to continue working on this rule, but I am will need some guidance as this is non-obvious to me. It feels much like I am coding in the dark, unable to easily debug what is happening when my rule is run (as well as how the infrastructure that is used to run the tests in this codebase actually works).

sindresorhus commented:
> They are currently failing in CI due to some weird module error: 'import' and 'export' may appear only with 'sourceType: module' (see here).

You need to rename the file extension to `.mjs`.

sindresorhus commented:
@fisker can probably give you some tips on debugging `fix`.

fisker commented:
You can disable it run on codebase by adding it [here](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/c5a43459e881d36e1aa79a7e41a0ec0a05b7e7b7/test/run-rules-on-codebase/lint.mjs#L23), we can enable it when `xo` updated.

fisker commented:
> It feels much like I am coding in the dark

I'm sorry it makes you feel bad, but `run-rules-on-codebase` is not a required step during implementing, it only need to be done before merge, mentioned in last step here https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/new-rule.md

sindresorhus commented:
@himynameisdave Still interested in finishing this? :)

himynameisdave commented:
@sindresorhus yep I'm still interested, although happy to hand this off to someone else if they have more time to work on it. 👍

sindresorhus commented:
Closing this to let someone else be able to continue it: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/971#issuecomment-933228408
